File: ppl-meta-frontend/lib/models/storage_location.dart
import 'package:json_annotation/json_annotation.dart';

part 'storage_location.g.dart';

/// Storage location model
@JsonSerializable()
class StorageLocation {
  @JsonKey(name: 'uuid')
  final String uuid;

  @JsonKey(name: 'user_id')
  final String userId;

  @JsonKey(name: 'name')
  final String name;

  @JsonKey(name: 'location_type')
  final String locationType;

  @JsonKey(name: 'base_path')
  final String basePath;

  @JsonKey(name: 'tier')
  final String tier;

  @JsonKey(name: 'is_active')
  final bool isActive;

  @JsonKey(name: 'is_default')
  final bool isDefault;

  @JsonKey(name: 'total_capacity_bytes')
  final int? totalCapacityBytes;

  @JsonKey(name: 'used_bytes')
  final int usedBytes;

  @JsonKey(name: 'file_count')
  final int fileCount;

  @JsonKey(name: 'usage_percentage')
  final double usagePercentage;

  @JsonKey(name: 'used_gb')
  final double usedGb;

  @JsonKey(name: 'total_capacity_gb')
  final double? totalCapacityGb;

  @JsonKey(name: 'free_gb')
  final double? freeGb;

  @JsonKey(name: 'mount_verified')
  final bool mountVerified;

  @JsonKey(name: 'last_verified_at')
  final String? lastVerifiedAt;

  @JsonKey(name: 'created_at')
  final String? createdAt;

  @JsonKey(name: 'updated_at')
  final String? updatedAt;

  StorageLocation({
    required this.uuid,
    required this.userId,
    required this.name,
    required this.locationType,
    required this.basePath,
    required this.tier,
    required this.isActive,
    required this.isDefault,
    this.totalCapacityBytes,
    this.usedBytes = 0,
    this.fileCount = 0,
    this.usagePercentage = 0.0,
    this.usedGb = 0.0,
    this.totalCapacityGb,
    this.freeGb,
    this.mountVerified = false,
    this.lastVerifiedAt,
    this.createdAt,
    this.updatedAt,
  });

  factory StorageLocation.fromJson(Map<String, dynamic> json) =>
      _$StorageLocationFromJson(json);
  Map<String, dynamic> toJson() => _$StorageLocationToJson(this);

  bool get isCloud =>
      locationType == 'cloud_s3' ||
      locationType == 'cloud_azure' ||
      locationType == 'cloud_gcp';

  bool get isUnlimited => totalCapacityBytes == null;

  String get typeLabel {
    switch (locationType) {
      case 'local_disk':
        return 'Local Disk';
      case 'external_drive':
        return 'External Drive';
      case 'cloud_s3':
        return 'AWS S3';
      case 'cloud_azure':
        return 'Azure Blob';
      case 'cloud_gcp':
        return 'Google Cloud';
      default:
        return locationType;
    }
  }

  String get typeIcon {
    switch (locationType) {
      case 'local_disk':
        return '💾';
      case 'external_drive':
        return '🔌';
      case 'cloud_s3':
      case 'cloud_azure':
      case 'cloud_gcp':
        return '☁️';
      default:
        return '📁';
    }
  }
}

/// Storage dashboard summary model
@JsonSerializable()
class StorageDashboard {
  @JsonKey(name: 'total_capacity_bytes')
  final int? totalCapacityBytes;

  @JsonKey(name: 'total_used_bytes')
  final int totalUsedBytes;

  @JsonKey(name: 'total_files')
  final int totalFiles;

  @JsonKey(name: 'usage_percentage')
  final double usagePercentage;

  @JsonKey(name: 'active_used_bytes')
  final int activeUsedBytes;

  @JsonKey(name: 'active_files')
  final int activeFiles;

  @JsonKey(name: 'archive_used_bytes')
  final int archiveUsedBytes;

  @JsonKey(name: 'archive_files')
  final int archiveFiles;

  @JsonKey(name: 'total_capacity_gb')
  final double? totalCapacityGb;

  @JsonKey(name: 'total_used_gb')
  final double totalUsedGb;

  @JsonKey(name: 'free_gb')
  final double? freeGb;

  @JsonKey(name: 'location_count')
  final int locationCount;

  @JsonKey(name: 'locations')
  final List<dynamic> locations;

  @JsonKey(name: 'alerts')
  final List<dynamic> alerts;

  @JsonKey(name: 'media_real_used_bytes')
  final int mediaRealUsedBytes;

  @JsonKey(name: 'media_real_files')
  final int mediaRealFiles;

  @JsonKey(name: 'media_real_used_gb')
  final double mediaRealUsedGb;

  @JsonKey(name: 'default_active_location')
  final Map<String, dynamic>? defaultActiveLocation;

  StorageDashboard({
    this.totalCapacityBytes,
    this.totalUsedBytes = 0,
    this.totalFiles = 0,
    this.usagePercentage = 0.0,
    this.activeUsedBytes = 0,
    this.activeFiles = 0,
    this.archiveUsedBytes = 0,
    this.archiveFiles = 0,
    this.totalCapacityGb,
    this.totalUsedGb = 0.0,
    this.freeGb,
    this.locationCount = 0,
    this.locations = const [],
    this.alerts = const [],
    this.mediaRealUsedBytes = 0,
    this.mediaRealFiles = 0,
    this.mediaRealUsedGb = 0.0,
    this.defaultActiveLocation,
  });

  factory StorageDashboard.fromJson(Map<String, dynamic> json) =>
      _$StorageDashboardFromJson(json);
  Map<String, dynamic> toJson() => _$StorageDashboardToJson(this);
}

/// Storage location create/update request
class StorageLocationRequest {
  final String name;
  final String locationType;
  final String basePath;
  final String tier;
  final bool isDefault;
  final Map<String, dynamic>? cloudConfig;

  StorageLocationRequest({
    required this.name,
    required this.locationType,
    required this.basePath,
    this.tier = 'active',
    this.isDefault = false,
    this.cloudConfig,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'location_type': locationType,
        'base_path': basePath,
        'tier': tier,
        'is_default': isDefault,
        if (cloudConfig != null) 'cloud_config': cloudConfig,
      };
}

File: ppl-meta-media/src/models/storage_location.py
"""
Storage Location Model

Defines user-configurable storage locations where media files can reside.
Supports local disk, external drives, and cloud storage providers.
"""

import uuid
from enum import Enum

from sqlalchemy import Boolean, BigInteger, Column, DateTime, Float, Integer, String, Text
from sqlalchemy import Enum as SQLEnum
from sqlalchemy import JSON
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func

from .base import BaseModel


class LocationType(Enum):
    """Storage location types."""

    LOCAL_DISK = "local_disk"
    EXTERNAL_DRIVE = "external_drive"
    CLOUD_S3 = "cloud_s3"
    CLOUD_AZURE = "cloud_azure"
    CLOUD_GCP = "cloud_gcp"


class StorageTier(Enum):
    """Storage tier classification."""

    ACTIVE = "active"
    ARCHIVE = "archive"


class StorageLocation(BaseModel):
    """
    User-defined storage location where media files can reside.

    Supports local disk paths, external/mounted drives, and cloud storage buckets.
    Each location tracks its capacity, usage, and availability status.
    """

    __tablename__ = "storage_locations"

    uuid = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, index=True)

    # Owner
    user_id = Column(UUID(as_uuid=True), nullable=False, index=True)

    # Display
    name = Column(String(255), nullable=False)  # e.g. "Main SSD", "NAS Backup"

    # Type and path
    location_type = Column(SQLEnum(LocationType), nullable=False)
    base_path = Column(
        String(1000), nullable=False
    )  # /Volumes/MediaDrive or s3://bucket

    # Status
    is_active = Column(Boolean, default=True, nullable=False)
    is_default = Column(Boolean, default=False, nullable=False)

    # Tier
    tier = Column(SQLEnum(StorageTier), default=StorageTier.ACTIVE, nullable=False)

    # Capacity tracking
    total_capacity_bytes = Column(BigInteger, nullable=True)  # null = unlimited (cloud)
    used_bytes = Column(BigInteger, default=0, nullable=False)
    file_count = Column(Integer, default=0, nullable=False)

    # Cloud credentials (encrypted JSON, null for local/external)
    cloud_config = Column(JSON, nullable=True)

    # Verification
    mount_verified = Column(Boolean, default=False, nullable=False)
    last_verified_at = Column(DateTime(timezone=True), nullable=True)
    last_scanned_at = Column(DateTime(timezone=True), nullable=True)

    @property
    def free_bytes(self):
        """Available bytes. None if unlimited."""
        if self.total_capacity_bytes is None:
            return None
        return max(0, self.total_capacity_bytes - self.used_bytes)

    @property
    def usage_percentage(self):
        """Usage as a percentage. 0 if unlimited."""
        if not self.total_capacity_bytes or self.total_capacity_bytes == 0:
            return 0.0
        return round((self.used_bytes / self.total_capacity_bytes) * 100, 2)

    @property
    def total_capacity_gb(self):
        """Total capacity in GB. None if unlimited."""
        if self.total_capacity_bytes is None:
            return None
        return round(self.total_capacity_bytes / (1024**3), 2)

    @property
    def used_gb(self):
        """Used space in GB."""
        return round(self.used_bytes / (1024**3), 2)

    @property
    def free_gb(self):
        """Free space in GB. None if unlimited."""
        if self.free_bytes is None:
            return None
        return round(self.free_bytes / (1024**3), 2)

    @property
    def uri_scheme(self):
        """URI scheme for this location type."""
        schemes = {
            LocationType.LOCAL_DISK: "file://",
            LocationType.EXTERNAL_DRIVE: "file://",
            LocationType.CLOUD_S3: "s3://",
            LocationType.CLOUD_AZURE: "azure://",
            LocationType.CLOUD_GCP: "gs://",
        }
        return schemes.get(self.location_type, "file://")

    @property
    def is_cloud(self):
        """Whether this is a cloud storage location."""
        return self.location_type in (
            LocationType.CLOUD_S3,
            LocationType.CLOUD_AZURE,
            LocationType.CLOUD_GCP,
        )

    def __repr__(self):
        return (
            f"<StorageLocation(uuid={self.uuid}, name='{self.name}', "
            f"type={self.location_type.value}, tier={self.tier.value})>"
        )
